src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    /// <summary>
+    /// Middleware that logs HTTP requests and HTTP responses.
+    /// </summary>
+    internal sealed class HttpLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+        private IOptionsMonitor<HttpLoggingOptions> _options;
+        private const int DefaultRequestFieldsMinusHeaders = 7;
+        private const int DefaultResponseFieldsMinusHeaders = 2;
+
+        /// <summary>
+        /// Initializes <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        /// <param name="loggerFactory"></param>
+        public HttpLoggingMiddleware(RequestDelegate next, IOptionsMonitor<HttpLoggingOptions> options, ILoggerFactory loggerFactory)
+        {
+            _next = next ?? throw new ArgumentNullException(nameof(next));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _options = options;
+            _logger = loggerFactory.CreateLogger<HttpLoggingMiddleware>();
+        }
+
+        /// <summary>
+        /// Invokes the <see cref="HttpLoggingMiddleware" />.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>HttpResponseLog.cs
+        public async Task Invoke(HttpContext context)
+        {
+            var options = _options.CurrentValue;
+            RequestBufferingStream? requestBufferingStream = null;
+            Stream? originalBody = null;
+            if ((HttpLoggingFields.Request & options.LoggingFields) != HttpLoggingFields.None)
+            {
+                var request = context.Request;
+                var list = new List<KeyValuePair<string, object?>>(
```

Comment:
nit: why is this `string, object?`, if I saw correctly, we're always using strings as values.

Reply:
Doesn't HttpRequestLog need to be `IReadOnlyList<KeyValuePair<string, object?>>`?

Reply:
I think it's TState, so should be fine

Reply:
It didn't work for me when I tried. I'll try again though.

Reply:
It looks like you're still adding some non-strings like `StatusCode` as the value. Not sure if it makes sense to convert this to string ourselves.

Should we be pooling the List?

Nit: I also think it would read better to change `KeyValuePair<string, object?>` to `(string, object?)` or `(string, string)` or something. Now that it's not public, we might as well get rid of all the extra `<` and `>` 😆

Reply:
https://github.com/dotnet/aspnetcore/issues/31961

Reply:
I'm going to convert to string for now on everything.